### DISCUSSION TITLE
What are the differences between neutralino and tauri?

### DISCUSSION THREAD

**Discussion by ghost** [neutralinojs/neutralinojs#542] [Q&A]
I've seen articles differentiating the two, but to a newcomer please understand that the differences are so small they seem trivial. Example:  I understand with maybe edge computing or small devices that 2MB vs 8MB bundle size might matter, but for desktops anything under 10MB is great anyway.

What are the biggest implications of these differences?:
- https://github.com/Elanis/web-to-desktop-framework-comparison
- https://news.ycombinator.com/item?id=23771558
- https://blog.logrocket.com/why-use-electron-alternative/

Also, FYI author claims bundle can be as low as < 1 MB with Tauri:
https://www.reddit.com/r/rust/comments/g9p2om/anybody_using_tauri/

**Comment by Erudition** [comment]
I'm wondering this too! So far it does seem like Tauri is more secure but more complicated, and I'd need to learn Rust to do anything fancy. On the other hand it's more likely to eventually support mobile.

https://blog.notesnook.com/neutralinojs-next-best-alternative-to-electron-and-tauri/

**Reply by Zireael07** [reply]
IIRC Neutralino won't support mobile

**Comment by Blatman** [comment]
Been looking at Tauri (and Wails - similar to Tauri but uses Go as backend) vs Neutralino and a major difference when developing is the fact Tauri ( & Wails) only supports a compiled binary.
In my case I use MacOS and quite like to have the app resources in the Resources folder as is quite normal - Neutralino allows this as well as supporting a bundling option using the resources.neu file. This means that you can use standard release binaries and then dev the HTML/JS/CSS resources without having to incessantly re-compile to test. With Tauri most changes require a re-compile. For those needing code protection/security then compilation is a requirement of course (but not necessarily during dev). I come from NW.js and the ability to dev the HTML/JS/CSS in the clear similar to Neutralino is great and with both the finished app can be compiled/packaged if required for distribution - I find dev with Tauri a lot slower because of the requirement to always compile.
(Neutralino - big win)
(I really hate incessant compilation during dev ... )

I also looked at Tauri to see if the webview features were more complete and indeed native menus are available - Tauri leverages the same webview that Neutralino does but I think there is more resource available to integrate the webview with Tauri than I think exists with Neutralino.
(Tauri - big win)

The frontend for both is JS but any backend work will need either C++ or Rust knowledge.
For me the key to desktop apps using HTML/JS/CSS is being able to have control of functions that are implicit within the OS such as native menus, dialogs, notifications, window management without having to get deeply into low-level languages. Both distros are trying to do this through API's. It seems that although all these distros are trying to do desktop with HTML/JS/CSS the web way of doing things seems to take over when things like native menus, window management, file management should just be there (ie: available to the frontend JS).
(Both have API's) Note: bugs in API's often need attention in low level upstream libraries where resourcing is a factor

Backend support for both distros requires more expertise from devs and I get the feeling that Tauri is ahead here with a bit more resource and attention. Neutralino uses a number of upstream libraries which could do with a lot more attention to bring robust and seamless functionality to the frontend via the backend code (eg: webview/native menus/dialogs). After initial implementation of the library ongoing improvement appears to stall.
(Tauri - win)

I have seen comment that Neutralino is highly insecure because it uses a localhost server but if your desktop app is a complete offline app is this really an issue? Tauri does have flexibility in this regard with the allowList feature which can limit access to various scopes.
Obviously any app with web/network access will require more focus on security and Tauri seems to be ahead there.
(Tauri - win)

Extensions/Plugins/Sidecars - both distros offer these but decent examples are sparse and the very minimal basic examples really depend on the dev being experienced with these concepts. The downside with Tauri apps is that the code can not be easily checked for learning unless the source is available.

As a side note I have noticed that most app examples available appear to be more web orientated than desktop so finding examples of code optimised for desktop is difficult.


I prefer Neutralino from the point of view of dev (no compilation required) but the backend/webview support by Tauri appears stronger and more robust (native menus/dialogs).

**Reply by CodeWithMufaddal** [reply]
soo what is the final conclusion ?
which to pick? why ?
as i have limited time to invest on any one tech i need to pick one?

**Reply by lukasMega** [reply]
Tauri is more mature, more tested project with much bigger community and better examples and documentation. 
Worth to mention the audits that was done and based on them lot of security issues was fixed.

But for simple apps Neutralino is good choice. You can very quickly get executable output ("compiled" binary).
